src/Datastore/DatastoreClient.php

```diff
@@ -0,0 +1,907 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+namespace Google\Cloud\Datastore;
+
+use Google\Cloud\ClientTrait;
+use Google\Cloud\Datastore\Query\GqlQuery;
+use Google\Cloud\Datastore\Query\Query;
+use Google\Cloud\Datastore\Query\QueryBuilder;
+use Google\Cloud\Datastore\Query\QueryInterface;
+use Google\Cloud\Datastore\Connection\Rest;
+use InvalidArgumentException;
+
+/**
+ * Google Cloud Datastore client. Cloud Datastore is a highly-scalable NoSQL
+ * database for your applications.  Find more information at
+ * [Google Cloud Datastore docs](https://cloud.google.com/datastore/docs/).
+ */
+class DatastoreClient
+{
+    use ClientTrait;
+    use DatastoreTrait;
+
+    const FULL_CONTROL_SCOPE = 'https://www.googleapis.com/auth/datastore';
+
+    const DEFAULT_READ_CONSISTENCY = 'EVENTUAL';
+
+    /**
+     * @var ConnectionInterface
+     */
+    protected $connection;
+
+    /**
+     * Create a Datastore client.
+     *
+     * Example:
+     * ```
+     * use Google\Cloud\ServiceBuilder;
+     *
+     * $cloud = new ServiceBuilder([
+     *     'projectId' => 'my-awesome-project'
+     * ]);
+     *
+     * $datastore = $cloud->datastore();
+     * ```
+     *
+     * ```
+     * // The Datastore client can also be instantianted directly.
+     * use Google\Cloud\Datastore\DatastoreClient;
+     *
+     * $datastore = new DatastoreClient([
+     *     'projectId' => 'my-awesome-project'
+     * ]);
+     * ```
+     *
+     * @param array $config {
+     *     Configuration Options.
+     *
+     *     @type string $projectId The project ID from the Google Developer's
+     *           Console.
+     *     @type callable $authHttpHandler A handler used to deliver Psr7
+     *           requests specifically for authentication.
+     *     @type callable $httpHandler A handler used to deliver Psr7 requests.
+     *     @type string $keyFile The contents of the service account
+     *           credentials .json file retrieved from the Google Developers
+     *           Console.
+     *     @type string $keyFilePath The full path to your service account
+     *           credentials .json file retrieved from the Google Developers
+     *           Console.
+     *     @type int $retries Number of retries for a failed request. Defaults
+     *           to 3.
+     *     @type array $scopes Scopes to be used for the request.
+     * }
+     * @throws \InvalidArgumentException
+     */
+    public function __construct(array $config = [])
+    {
+        if (!isset($config['scopes'])) {
+            $config['scopes'] = [self::FULL_CONTROL_SCOPE];
+        }
+
+        $this->connection = new Rest($this->configureAuthentication($config));
+    }
+
+    /**
+     * Create a single Key instance
```

Comment:
A little detail about how we detect whether to use id/name here seems like it could be very helpful.